src/Algebra/Module/Construct/Nagata.agda

```diff
@@ -0,0 +1,170 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- The non-commutative analogue of Nagata's construction
+-- of the "idealization of a module", defined here
+-- on R-R-*bi*modules M over a ring R.
+--
+-- Elements of R ⋉ M are pairs |R| × |M|
+------------------------------------------------------------------------
+
+{-# OPTIONS --cubical-compatible --safe #-}
+
+module Algebra.Module.Construct.Nagata where
+
+open import Algebra.Bundles
+open import Algebra.Core
+import Algebra.Consequences.Setoid as Consequences
+import Algebra.Definitions as Definitions
+open import Algebra.Module.Bundles
+open import Algebra.Module.Core
+open import Algebra.Module.Definitions
+import Algebra.Module.Construct.DirectProduct as DirectProduct
+import Algebra.Module.Construct.TensorUnit as TensorUnit
+open import Algebra.Structures
+open import Data.Product using (_×_; _,_; proj₁; proj₂)
+open import Data.Product.Relation.Binary.Pointwise.NonDependent
+open import Level using (Level; _⊔_)
+open import Relation.Binary using (Rel; Setoid; IsEquivalence)
+import Relation.Binary.Reasoning.Setoid as ≈-Reasoning
+
+private
+  variable
+    m ℓm r ℓr : Level
+
+------------------------------------------------------------------------
+-- Definitions
+
+module Nagata (ring : Ring r ℓr) (bimodule : Bimodule ring ring m ℓm) where
+
+  open Ring ring
+    renaming (Carrier to R
+             ; 0# to 0ᴿ
+             ; 1# to 1ᴿ
+             ; _+_ to _+ᴿ_
+             ; _*_ to _*ᴿ_
+             ; *-cong to *ᴿ-cong
+             ; *-assoc to *ᴿ-assoc
+             ; *-identityʳ to *ᴿ-identityʳ
+             ; *-identityˡ to *ᴿ-identityˡ
+             ; distribˡ to *ᴿ-distribˡ-+ᴿ
+             ; distribʳ to *ᴿ-distribʳ-+ᴿ
+             )
+
+  open Bimodule bimodule
+    renaming (Carrierᴹ to M
+             )
+
+  open AbelianGroup +ᴹ-abelianGroup
+    renaming (setoid to setoidᴹ
+             ; isEquivalence to isEquivalenceᴹ
+             )
+
+  open IsEquivalence isEquivalenceᴹ
+    renaming (sym to symᴹ)
+
+  +ᴺ-bimodule = DirectProduct.bimodule TensorUnit.bimodule bimodule
+
+  open Bimodule +ᴺ-bimodule public using ()
+    renaming (Carrierᴹ to N
+             ; _≈ᴹ_ to _≈ᴺ_
+             ; _+ᴹ_ to _+ᴺ_
+             ; 0ᴹ to 0ᴺ
+             ; -ᴹ_ to -ᴺ_
+             ; +ᴹ-abelianGroup to +ᴺ-abelianGroup
+             )
+
+  ιᴿ : R → N
+  ιᴿ r = r , 0ᴹ
+
+  ιᴹ : M → N
+  ιᴹ m = 0ᴿ , m
+
+  1ᴺ : N
+  1ᴺ = ιᴿ 1ᴿ
+
+  open AbelianGroup +ᴺ-abelianGroup
```

Comment:
Personally, I prefer all my `open` together and not interleave them with definitions.

Reply:
Yes, OK, have (hopefully, now) done so. Some `open`s do require intermediate defns though, such as that of the `+-bimodule` structure...